src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UserReleaseInviteRepository.cs

```diff
@@ -0,0 +1,120 @@
+﻿#nullable enable
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using GovUk.Education.ExploreEducationStatistics.Admin.Services.Interfaces;
+using GovUk.Education.ExploreEducationStatistics.Common.Model;
+using GovUk.Education.ExploreEducationStatistics.Content.Model;
+using GovUk.Education.ExploreEducationStatistics.Content.Model.Database;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using Guid = System.Guid;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Services
+{
+    public class UserReleaseInviteRepository : IUserReleaseInviteRepository
+    {
+        private readonly ContentDbContext _contentDbContext;
+
+        public UserReleaseInviteRepository(ContentDbContext contentDbContext)
+        {
+            _contentDbContext = contentDbContext;
+        }
+
+        public async Task<Either<ActionResult, Unit>> Create(Guid releaseId, string email,
+            ReleaseRole releaseRole, bool emailSent, Guid createdById, bool accepted = false)
+        {
+            await _contentDbContext.AddAsync(
+                new UserReleaseInvite
+                {
+                    Email = email.ToLower(),
+                    ReleaseId = releaseId,
+                    Role = releaseRole,
+                    Accepted = accepted,
+                    EmailSent = emailSent,
+                    Created = DateTime.UtcNow,
+                    CreatedById = createdById,
+                }
+            );
+
+            await _contentDbContext.SaveChangesAsync();
+
+            return Unit.Instance;
+        }
+
+        public async Task<Either<ActionResult, Unit>> CreateManyIfNotExists(List<Guid> releaseIds, string email,
+            ReleaseRole releaseRole, bool emailSent, Guid createdById, bool accepted = false)
+        {
+            foreach (var releaseId in releaseIds)
+            {
+                if (!await UserHasInvite(releaseId, email, releaseRole))
+                {
+                    await _contentDbContext.AddAsync(
```

Comment:
Not sure if it makes a big difference performance wise, but we could use `AddRangeAsync` to add many entities at the same time. In combination with `ToAsyncEnumerable` (one of the correct cases to use it), we could refactor it to something like this:

```cs
await _contentDbContext.AddRangeAsync(
    releaseIds.ToAsyncEnumerable()
        .WhereAwait(releaseId => UserHasInvite(releaseId, email, releaseRole))
        .Select(releaseId => ...)
);
```

Reply:
I've made the change. I did a cursory test of the above change and it seems to work fine.

I had originally written this code using async stuff - not exactly as you've suggested - and I was getting strange errors where not all the invites were getting created. So I switched to using the foreach, and that solved it.